crypto/dh/dh_key.c

```diff
@@ -27,16 +27,32 @@ int DH_generate_key(DH *dh)
 
 int DH_compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
 {
-    return dh->meth->compute_key(key, pub_key, dh);
+    int ret = 0;
+    BIGNUM *Z = NULL;
+
+    /* compute the key; ret is constant unless compute_key is external */
+    if ((ret = dh->meth->compute_key(key, pub_key, dh)) <= 0)
+        return ret;
+
+    /* reconstruct the BIGNUM from the key bytes */
+    if ((Z = BN_bin2bn(key, ret, NULL)) == NULL)
+        return -1;
+
+    /* strip the leading zeros -- i.e. unpad the key */
+    ret = BN_bn2bin(Z, key);
+    BN_clear_free(Z);
```

Comment:
I wonder whether removing the padding zeros here is really needed. IMO nothing should really depend on the DH_compute_key not returning any leading zeros.

Reply:
Then why do we have 2 function?

Reply:
The padded version was added in commit bc91494e064ebdcff68f987947f97e404fbca0b5, to be SP 800-56A compliant

Reply:
From the Raccoon paper, Sect 4:

> SSLv3, TLS 1.0, TLS 1.1, and TLS 1.2 prescribe stripping leading zero bytes of the computed DH shared secret.

I know that `make test` will barf all over the place without the zero stripping.

Reply:
To give a little more info with fresh eyes this morning.

> barf all over the place

Not exactly, but anyway. Before we started testing mitigations, we wanted to make sure regression tests were in place. And it turns out this is internally already being explicitly tested

https://github.com/openssl/openssl/blob/9be10637502bf32189055dff8d3442e140e845c5/test/recipes/30-test_evp_data/evppkey.txt#L17930-L17941

So the `make test` failure is predictable:

```
# ERROR: (memory) 'expected == got' failed @ test/evp_test.c:92
# --- expected
# +++ got
# 0000:-6620dd85b56ee854 0c8040cac46b7385 344a164e4dbdf521 f7d99f88fa68edd2
# 0000:+00006620dd85b56e e8540c8040cac46b 7385344a164e4dbd f521f7d99f88fa68
#       ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^ ^^^^^^^^^^^^^^^^
```

Yes you can argue "change the test" but I don't think the `DH_compute_key` behavior can be changed for 111 (this PR), and it would be a breaking change for master. (Although I would support such a change. Having these variable-length secrets floating around is a horrible idea. But it's a question of standards compatibility and existing OpenSSL behavior.)

Reply:
OK, then. But I'd still at least avoid doing BN conversion here and just count the initial 0 bytes and memmove the key over them.

Reply:
The thing is, we had that. But we felt it made the original Raccoon attack vulnerability worse.

Reply:
I can't really see how it could be worse than the conversion to BN and back to data. There will be memory copying based on the length of the pad-removed data anyway.

Reply:
conversion to BN will be CT because of this PR. But yes, then the question of "back to data"

Reply:
Distilling, everything should be addressed in a04d46a9b10a5f3f61f82fe992c4cf4393355dde force push